### PR TITLE
Calendar on ecliptic

### LINKED ISSUE #2175: Calendar on Ecliptic
Stellarium 0.21.2 - It would be useful to have a calendar with division into days on the Ecliptic, to show the positions of the sun during the months

### PR DESCRIPTION
**Pull request by rayw000**
### Description

Fixes #2175 

### Screenshots (if appropriate):

image

Help wanted:

Is there any function which I could use to get the ecliptic longitude of the sun on a specific Julian day? I didn't find it, so I divided the ecliptic equally into 365 (366 if leap year) pieces. As we known that the earth moves at different angular velocities in different seasons on ecliptic, equally dividing would cause bad precision. Sometimes there would be up to 2 days deviation.

Many thanks!

I've tried to use `earth->getEclipticPos(<JDE>)` to get the `Vec3d` position of earth, and calculate the differences of each adjacent days, and use them as the steps to split the ecliptic, but the differences are widely variable. I possibly misunderstand something.

### Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] This change requires a documentation update

### How Has This Been Tested?

**Test Configuration**: macOS Monterey 12.0.1
* Operating system: macOS Monterey 12.0.1
* Graphics Card: Apple M1 Max built-in

## Checklist:

- [x] My code follows the [code style](http://stellarium.org/doc/head/codingStyle.html) of this project.
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation (header file)
- [ ] I have updated the respective chapter in the Stellarium User Guide
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Comment by gzotti** [comment]
Rather use earth->getEclipticPos(JDE); for each consecutive day of the year (or use "today -182" to "today+183" or so)? This may become slow, though. Probably it will be needed to compute earth's position without the rest of the solar system. 

The ecliptic otherwise can be just marked in 360 regular degrees which allow a good estimate of motion.

**Comment by rayw000** [comment]
> Rather use earth->getEclipticPos(JDE); for each consecutive day of the year (or use "today -182" to "today+183" or so)? This may become slow, though. Probably it will be needed to compute earth's position without the rest of the solar system.
> 
> The ecliptic otherwise can be just marked in 360 regular degrees which allow a good estimate of motion.

Thank you @gzotti 

Marking the ecliptic in 360 regular degrees starting from Jan 1st is my current implementation. But due to variable angular velocity the earth moves at, there may be up to 2 days deviation from the position the sun should be in on the ecliptic.

How would you think if we mark the solstices and the equinoxes with dates, and divide each three months between them? This would give better precision than just equally dividing the whole ecliptic.

**Comment by rayw000** [comment]
I pushed a new commit. I divided the half year from autumnal equinox to spring equinox into 178 pieces, while spring equinox to autumnal equinox into 187 (188 if leap year) pieces. Now we have a better approximation, with deviations no more than 1 day.

**Comment by rayw000** [comment]
Feature's done.

This PR decreases the code coverage of file `GridLinesMgr.cpp` from `11.458%` to `11.452%` (60 new lines). I didn't see function `void SkyLine::draw(StelCore*) const` called in any unit tests. Should we ignore this failing check?

**Comment by alex-w** [comment]
> Should we ignore this failing check?

Yes

**Comment by alex-w** [comment]
A quick check: today 10 February, 19:00 UTC+7 - on the screenshot you may see end of 9 February:
stellarium-001

**Comment by rayw000** [comment]
> A quick check: today 10 February, 19:00 UTC+7 - on the screenshot you may see end of 9 February: stellarium-001

The left span of Feb 10 is Feb 10. The date label is at the end of the day, for example Feb 1 starts at the `Feb` mark and ends at the first short mark. It is the expected behavior.

File: src/core/modules/GridLinesMgr.cpp [hunk 1]
```diff
@@ -1229,6 +1229,107 @@ void SkyLine::draw(StelCore *core) const
...
+				};
+				for(int i = 0,d = 0; i < daysOfYear; ++i, ++d)
+				{
+					if(d >= monthDays)
+					{
+						d = 0;
+						++month;
+						monthDays = StelUtils::numberOfDaysInMonthInYear(month, year);
+						sPainter.drawGreatCircleArc(partNewYear, partMonth, Q_NULLPTR, Q_NULLPTR, Q_NULLPTR);
+						QString label = QString("%1").arg(monthNames[month - 1]);
```
[the remark this answers is not in the document]
Comment (rayw000) [reply src/core/modules/GridLinesMgr.cpp]: Done

**Comment by rayw000** [comment]
> I have not checked yet and can do that only later: What about GravityLabels, and what about long-time errors from assuming the perihel is always near (north) winter solstice? What happens in 1582 ;-) ? The optimal solution would of course draw strict day (0h) marks in observer local time (as suggested, -182...+183 from "now" (maybe with a cache of JD/values), but at what cost?

Perhaps drawing strict day marks from "now" would cause larger deviations at the other side of the ecliptic?

**Comment by alex-w** [comment]
@gzotti you are right, I see strange results in the past!

**Comment by rayw000** [comment]
Splitting into 178+187 pieces is not precise, but just cheap.

**Comment by rayw000** [comment]
I've tried to get the `Vec3d` position `p` of earth and calculate the differences between each two consequent `acos(p[0])`. It has bad precision near 0 and Pi.

**Comment by gzotti** [comment]
> Perhaps drawing strict day marks from "now" would cause larger deviations at the other side of the ecliptic?

The opposite side of the ecliptic is of course where the open ends will fail to connect visually smoothly. There will be two marks 1/4 or 5/4 degree apart, and the pair will run in-sync with the sun. However, it is the least-interesting point to look at on this scale. 

Older installations, like the optomechanical planetarium, of course have to use some "average" scale, usually showing days 2 years past leap years to keep the error minimized. But computers can do better :-)

To create the lines accurately, the solar position should be computed not by setting StelCore->setJD(), doing all the heavy computations for all objects, but with the underlying functions. If this is left as exercise to myself, we can accept this now and improve it later. However, are GravityLabels solved?

**Comment by gzotti** [comment]
> I've tried to get the `Vec3d` position `p` of earth and calculate the differences between each two consequent `acos(p[0])`. It has bad precision near 0 and Pi.

Not sure what you mean. You would need earth's position, and derive Solar position as -earth. But 365 times on 365 dates per full circle. Therefore I recommend caching the results, so as long as you don't simulate millennia at a time, the step coordinates will mostly be re-used.

**Comment by gzotti** [comment]
> > However, are GravityLabels solved?
> 
> If I read correct, do you mean we should make the ecliptic labels behave like constellation names?

Yes. Switch on the planetarium distortion mode in the settings menu and see how all texts are now oriented so that they would be readable when projected onto a planetarium dome.

**Comment by rayw000** [comment]
> Not sure what you mean. You would need earth's position, and derive Solar position as -earth.

Yes, that's what I've tried. This should give us only 365 positions for each 0h, e.g. [-0.999935, y, z]. The first element could range from -0.999xxx to 1.03xxx. I think their `acos` value could be used as the coordinates on the ecliptic.

**Comment by rayw000** [comment]
> Yes. Switch on the planetarium distortion mode in the settings menu and see how all texts are now oriented so that they would be readable when projected onto a planetarium dome.

It could be done by passing fixed values to `StelPainter::drawText`. Am I right?

**Comment by rayw000** [comment]
> > Yes. Switch on the planetarium distortion mode in the settings menu and see how all texts are now oriented so that they would be readable when projected onto a planetarium dome.
> 
> It could be done by passing fixed values to `StelPainter::drawText`. Am I right?

Oh, you mean it should be responsive to the settings menu. I'll do some research.

**Comment by rayw000** [comment]
image

> Yes. Switch on the planetarium distortion mode in the settings menu and see how all texts are now oriented so that they would be readable when projected onto a planetarium dome.

I know what you mean.lol

**Comment by gzotti** [comment]
> > Not sure what you mean. You would need earth's position, and derive Solar position as -earth.
> 
> Yes, that's what I've tried. This should give us only 365 positions for each 0h, e.g. [-0.999935, y, z]. The first element could range from -0.999xxx to 1.03xxx. I think their (normalized) `acos` value could be used as the coordinates on the ecliptic.

Not so fast... These are J2000 rectangular (Cartesian) ecliptical coordinates. We have to apply precession/nutation/aberration to current date to make them usable for the "ecliptic of date".

**Review by alex-w — CHANGES_REQUESTED** [review]
> The right span of 10 is Feb 10. The date label is at the end of the day, for example Feb 1 starts at the Feb mark and ends at the first short mark. It is the expected behavior.

This is a little weird. I’m checked hard copies of celestial maps and I didn’t find similar markings. This behaviour will confusing users.

**Comment by rayw000** [comment]
> > The right span of 10 is Feb 10. The date label is at the end of the day, for example Feb 1 starts at the Feb mark and ends at the first short mark. It is the expected behavior.
> 
> This is a little weird. I’m checked hard copies of celestial maps and I didn’t find similar markings. This behaviour will confusing users.

Or, we can put the label on the top middle of each span, something like

```
      3           2          1      Feb
|-----------|-----------|-----------|

```

**Comment by rayw000** [comment]
> This is a little weird. I’m checked hard copies of celestial maps and I didn’t find similar markings. This behaviour will confusing users.

@alex-w Done.

**Comment by gzotti** [comment]
We have found an accurate solution (#3057), but thank you for your interest!

**Pull request closed without merging**